pkg/querier/block_streaming.go

```diff
@@ -0,0 +1,280 @@
+}
+
+func (bqs *blockStreamingQuerierSeries) Iterator(reuse chunkenc.Iterator) chunkenc.Iterator {
+	// Fetch the chunks from the stream.
+	var allChunks []storepb.AggrChunk
+	for i := bqs.seriesIdxStart; i <= bqs.seriesIdxEnd; i++ {
+		chks, err := bqs.streamReader.GetChunks(uint64(i))
+		if err != nil {
+			return series.NewErrIterator(err)
+		}
+		allChunks = append(allChunks, chks...)
+	}
+	if len(allChunks) == 0 {
+		// should not happen in practice, but we have a unit test for it
+		return series.NewErrIterator(errors.New("no chunks"))
+	}
+
+	sort.Slice(allChunks, func(i, j int) bool {
+		return allChunks[i].MinTime < allChunks[j].MinTime
+	})
+
+	it, err := newBlockQuerierSeriesIterator(reuse, bqs.Labels(), allChunks)
+	if err != nil {
+		return series.NewErrIterator(err)
+	}
+
+	return it
+}
+
+// storeGatewayStreamReader is responsible for managing the streaming of chunks from a storegateway and buffering
+// chunks in memory until they are consumed by the PromQL engine.
+type storeGatewayStreamReader struct {
+	client              storegatewaypb.StoreGateway_SeriesClient
+	expectedSeriesCount int
+	queryLimiter        *limiter.QueryLimiter
+	stats               *stats.Stats
+	log                 log.Logger
+
+	seriesChunksChan chan *storepb.StreamingChunksBatch
+	chunksBatch      []*storepb.StreamingChunks
+	errorChan        chan error
+}
+
+func newStoreGatewayStreamReader(client storegatewaypb.StoreGateway_SeriesClient, expectedSeriesCount int, queryLimiter *limiter.QueryLimiter, stats *stats.Stats, log log.Logger) *storeGatewayStreamReader {
+	return &storeGatewayStreamReader{
+		client:              client,
+		expectedSeriesCount: expectedSeriesCount,
+		queryLimiter:        queryLimiter,
+		stats:               stats,
+		log:                 log,
+		seriesChunksChan:    make(chan *storepb.StreamingChunksBatch, 1),
+		// Important: to ensure that the goroutine does not become blocked and leak, the goroutine must only ever write to errorChan at most once.
+		errorChan: make(chan error, 1),
+	}
+}
+
+// Close cleans up all resources associated with this storeGatewayStreamReader.
+// This method should only be called if StartBuffering is not called.
+func (s *storeGatewayStreamReader) Close() {
+	if err := s.client.CloseSend(); err != nil {
+		level.Warn(s.log).Log("msg", "closing storegateway client stream failed", "err", err)
+	}
+}
+
+// StartBuffering begins streaming series' chunks from the storegateway associated with
+// this storeGatewayStreamReader. Once all series have been consumed with GetChunks, all resources
+// associated with this storeGatewayStreamReader are cleaned up.
+// If an error occurs while streaming, a subsequent call to GetChunks will return an error.
+// To cancel buffering, cancel the context associated with this storeGatewayStreamReader's storegatewaypb.StoreGateway_SeriesClient.
+func (s *storeGatewayStreamReader) StartBuffering() {
+	ctxDone := s.client.Context().Done()
+
+	go func() {
```

Comment:
Might be useful to add a trace span for this, similar to what was added for ingester streaming in #5315 (https://github.com/grafana/mimir/pull/5315/files#diff-75adda2fac178c8577513b34e897ddf223ad874fd5376ef39d970bf2bf594891)